dot/core/service.go

```diff
@@ -579,3 +611,41 @@ func (s *Service) GetMetadata(bhash *common.Hash) ([]byte, error) {
 	s.rt.SetContextStorage(ts)
 	return s.rt.Metadata()
 }
+
+// RegisterRuntimeUpdatedChannel function to register chan that is notified when runtime version changes
+func (s *Service) RegisterRuntimeUpdatedChannel(ch chan<- runtime.Version) (byte, error) {
+	s.runtimeUpdateSubscriptionsLock.Lock()
+	defer s.runtimeUpdateSubscriptionsLock.Unlock()
+
+	if len(s.runtimeUpdateSubscriptions) == 256 {
+		return 0, errors.New("channel limit reached")
+	}
+
+	var id byte
+	for {
+		id = generateID()
+		if s.runtimeUpdateSubscriptions[id] == nil {
+			break
+		}
+	}
+
+	s.runtimeUpdateSubscriptions[id] = ch
+	return id, nil
+}
+
+// UnregisterRuntimeUpdatedChannel function to unregister runtime updated channel
+func (s *Service) UnregisterRuntimeUpdatedChannel(id byte) bool {
+	ch := s.runtimeUpdateSubscriptions[id]
```

Comment:
check if the id exists
```suggestion
	ch, ok := s.runtimeUpdateSubscriptions[id]
```

Reply:
updated.

Reply:
@timwu20 I created `TestService_RegisterUpRegisterPanic` to test parallel calls, and the test panics (although I'm not sure if it's for the correct reason).  Please take a look and let me know if I'm on the right track.

Reply:
What panic did you get?

Reply:
I pulled your branch and I'm not getting any panics

Reply:
Currently there is a `t.Skip()` on the test TestService_RegisterUnRegisterPanic since it fails with a fatal error: concurrent map read and map write, so not the panics were expecting?

Reply:
oh well that should be fixed too.  Let me try and reproduce on my end.